web/packages/teleport/src/DesktopSession/Withholder.tsx

```diff
@@ -0,0 +1,123 @@
+ * the event to prevent the remote server from mistakenly thinking the key is held down when the user returns to the browser window, which can
+ * cause issues as described in https://github.com/gravitational/teleport/issues/24342.
+ * - For alt/cmd UP events, it introduces a short delay before handling to avoid unintended actions (like opening the start menu), in the case
+ * that an alt/cmd + tab registers both the alt/cmd DOWN and UP events in quick succession. (This can happen if the user does an alt/cmd + tab
+ * really quickly, according to our testing.)
+ * - For other keys, it handles them immediately.
+ *
+ * Events are either processed immediately, delayed, or cancelled based on user actions and focus changes.
+ */
+export class Withholder {
+  /**
+   * The list of keys which are to be withheld.
+   */
+  private keysToWithhold: string[] = ['Meta', 'Alt'];
+  /**
+   * The internal map of keystrokes that are currently
+   * being withheld.
+   */
+  private withheldKeys: Array<WithheldKeyboardEventHandler> = [];
+
+  /**
+   * All keyboard events should be handled via this function.
+   */
+  public handleKeyboardEvent(
+    params: KeyboardEventParams,
+    handleKeyboardEvent: (params: KeyboardEventParams) => void
+  ) {
+    const key = params.e.key;
+
+    // If this is not a key we withhold, immediately flush any withheld keys
+    // and handle this key.
+    if (!this.keysToWithhold.includes(key)) {
+      this.flush();
+      handleKeyboardEvent(params);
+      return;
+    }
+
+    // This is a key we withhold:
+
+    // On key down we withhold without a timeout. The handler will ultimately be called
+    // when the key is released (typically after a timeout, see the comment in the
+    // conditional below), or when another key is pressed, or else it should be cancelled
+    // onfocusout or on unmount.
+    let timeout = undefined;
+    if (params.state === ButtonState.UP) {
+      // On key ups we withhold on a timeout. The function will be called when the
+      // timer times out, or when another key is pressed, or else it should be
+      // cancelled onfocusout or on unmount.
+      timeout = setTimeout(() => {
+        // Just flush after the timeout, the handler will
+        // be in the queue by then and thus will be called.
+        //
+        // Technically this might flush some keys that were
+        // pressed after this one, but that works okay in practice.
+        // A user would have to be doing something extremely unusual
+        // for this to become a noticeable problem.
+        this.flush();
+      }, 10); // 10 ms was determined empirically to work well.
+    }
+
+    this.withheldKeys.push({
+      params,
+      handler: handleKeyboardEvent,
+      timeout,
+    });
+  }
+
+  // Cancel all withheld keys.
+  public cancel() {
+    while (this.withheldKeys.length > 0) {
+      const withheld = this.withheldKeys.shift();
```

Comment:
I suspect this is less efficient than something like:

```
this.withheldKeys.forEach(w => clearTimeout(w.timeout));
this.withheldKeys = [];
```

Take it or leave it. Probably doesn't matter much, especially when the array is small.